### PR TITLE
[Merged by Bors] - feat(Probability/Kernel/Disintegration): Uniqueness of the disintegration kernel

### PR DESCRIPTION
**Pull request by kex-y**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by RemyDegenne** [comment]
Can you shorten the proof of the real case using the pi-systems of #6490 instead of using `Iic` over the reals?

File: Mathlib/MeasureTheory/Measure/MeasureSpaceDef.lean [hunk 1]
```diff
@@ -559,6 +565,17 @@ theorem inter_ae_eq_empty_of_ae_eq_empty_right (h : t =ᵐ[μ] (∅ : Set α)) :
   rw [inter_empty]
 #align measure_theory.inter_ae_eq_empty_of_ae_eq_empty_right MeasureTheory.inter_ae_eq_empty_of_ae_eq_empty_right
 
+theorem _root_.MeasurableSpace.ae_induction_on_inter {β} [MeasurableSpace β] {μ : Measure β}
```
Comment (sgouezel) [inline Mathlib/MeasureTheory/Measure/MeasureSpaceDef.lean:568]: Could you add a docstring to this one?

File: Mathlib/MeasureTheory/Measure/MeasureSpaceDef.lean [hunk 2]
```diff
@@ -423,6 +423,12 @@ theorem ae_all_iff {ι : Sort*} [Countable ι] {p : α → ι → Prop} :
   eventually_countable_forall
 #align measure_theory.ae_all_iff MeasureTheory.ae_all_iff
 
+theorem all_ae_of {ι : Sort _} {p : α → ι → Prop} (hp : ∀ᵐ a ∂μ, ∀ i, p a i) (i : ι) :
+    ∀ᵐ a ∂μ, p a i := by
+  change μ { a | ¬ ∀ i, p a i } = 0 at hp
+  simp_rw [not_forall, Set.setOf_exists fun i x => ¬p x i] at hp
+  exact le_zero_iff.1 <| hp ▸ measure_mono (Set.subset_iUnion (fun i => {a | ¬p a i}) i)
```
Comment (sgouezel) [inline Mathlib/MeasureTheory/Measure/MeasureSpaceDef.lean:430]:
```suggestion
  filter_upwards [hp] with a ha using ha i
```

File: Mathlib/Probability/Kernel/Disintegration.lean [hunk 3]
```diff
@@ -477,6 +478,121 @@ theorem set_integral_condKernel_univ_left {ρ : Measure (α × Ω)} [IsFiniteMea
 
 end IntegralCondKernel
 
+section Unqiue
```
Comment (sgouezel) [inline Mathlib/Probability/Kernel/Disintegration.lean:481]:
```suggestion
section Unique
```

File: Mathlib/Probability/Kernel/Disintegration.lean [hunk 4]
```diff
@@ -477,6 +478,121 @@ theorem set_integral_condKernel_univ_left {ρ : Measure (α × Ω)} [IsFiniteMea
...
+    · rfl
+    · exact measurable_prod_mk_left hs
+  simp only [hprod, kernel.compProd_apply _ _ _ hs, kernel.prodMkLeft_apply,
+    kernel.map_apply _ hf.measurable, hinteq, Set.mem_preimage, Prod_map, id_eq,
+    kernel.lintegral_const]
+  rw [Measure.map_apply (measurable_id.prod_map hf.measurable) hs, ← lintegral_condKernel_mem]
+  · rfl
+  · exact measurable_id.prod_map hf.measurable hs
+
+end Unqiue
```
Comment (sgouezel) [inline Mathlib/Probability/Kernel/Disintegration.lean:594]:
```suggestion
end Unique
```

File: Mathlib/Probability/Kernel/CondDistrib.lean [hunk 5]
```diff
@@ -106,6 +106,21 @@ theorem aestronglyMeasurable'_integral_condDistrib (hX : AEMeasurable X μ) (hY
...
+    ∀ᵐ x ∂μ.map X, κ x = condDistrib Y X μ x := by
+  have heq : μ.map X = (μ.map (fun x => (X x, Y x))).fst
+  · ext s hs
+    rw [Measure.map_apply hX hs, Measure.fst_apply hs, Measure.map_apply]
+    exacts [rfl, Measurable.prod hX hY, measurable_fst hs]
+  rw [heq, condDistrib]
+  refine' eq_condKernel_of_measure_eq_compProd _ _ _
+  convert hκ
+  exact heq.symm
 section Integrability
```
Comment (RemyDegenne) [inline Mathlib/Probability/Kernel/CondDistrib.lean:124]:
```suggestion
  exact heq.symm

section Integrability
```

File: Mathlib/Probability/Kernel/Disintegration.lean [hunk 6]
```diff
@@ -477,6 +478,121 @@ theorem set_integral_condKernel_univ_left {ρ : Measure (α × Ω)} [IsFiniteMea
...
+      h _ <| hf.measurableSet_image.2 hs]
+  have hprod : (ρ.map (Prod.map id f)).fst = ρ.fst
+  · ext s hs
+    rw [Measure.fst_apply hs,
+      Measure.map_apply (measurable_id.prod_map hf.measurable) (measurable_fst hs),
+      ← Set.preimage_comp, Measure.fst_apply hs]
+    rfl
+  suffices : ρ.map (Prod.map id f) =
+    (kernel.const Unit (ρ.map (Prod.map id f)).fst ⊗ₖ
+     kernel.prodMkLeft Unit (kernel.map (Measure.condKernel ρ) f hf.measurable)) ()
```
Comment (RemyDegenne) [inline Mathlib/Probability/Kernel/Disintegration.lean:575]:
```suggestion
    (kernel.const Unit (ρ.map (Prod.map id f)).fst ⊗ₖ
      kernel.prodMkLeft Unit (kernel.map (Measure.condKernel ρ) f hf.measurable)) ()
```

**Review by RemyDegenne — APPROVED** [review]
Thanks!
bors d+

**Pull request closed without merging**